File: hw/cevallos/u3/hw29Singleton/python/views/sales_view.py
import tkinter as tk
from tkinter import ttk, messagebox
from typing import Callable, List, Dict, Any


class SalesView:
    """GUI view for sales management."""
    
    def __init__(self, root: tk.Tk) -> None:
        self.root = root
        self.root.title("Sales Management System")
        self.root.geometry("800x600")
        
        # Callbacks
        self.on_apply_discount_callback: Callable[[float], bool] = None
        self.on_create_sale_callback: Callable[[str, float], Any] = None
        
        self._setup_ui()
    
    def _setup_ui(self) -> None:
        """Setup the user interface."""
        # Main container
        main_frame = ttk.Frame(self.root, padding="10")
        main_frame.grid(row=0, column=0, sticky=(tk.W, tk.E, tk.N, tk.S))
        
        # Configure grid weights
        self.root.columnconfigure(0, weight=1)
        self.root.rowconfigure(0, weight=1)
        main_frame.columnconfigure(1, weight=1)
        
        # Discount control section
        discount_frame = ttk.LabelFrame(main_frame, text="Discount Control", padding="10")
        discount_frame.grid(row=0, column=0, columnspan=2, sticky=(tk.W, tk.E), pady=(0, 10))
        
        ttk.Label(discount_frame, text="Discount Percentage:").grid(row=0, column=0, padx=(0, 5))
        
        self.discount_var = tk.StringVar(value="10.0")
        self.discount_entry = ttk.Entry(discount_frame, textvariable=self.discount_var, width=15)
        self.discount_entry.grid(row=0, column=1, padx=(0, 10))
        
        ttk.Label(discount_frame, text="%").grid(row=0, column=2, padx=(0, 10))
        
        self.apply_discount_btn = ttk.Button(
            discount_frame, 
            text="Apply Discount",
            command=self._on_apply_discount
        )
        self.apply_discount_btn.grid(row=0, column=3)
        
        # Current discount display
        self.current_discount_label = ttk.Label(discount_frame, text="Current Discount: 10.0%")
        self.current_discount_label.grid(row=0, column=4, padx=(20, 0))
        
        # Sales input section
        input_frame = ttk.LabelFrame(main_frame, text="New Sale", padding="10")
        input_frame.grid(row=1, column=0, columnspan=2, sticky=(tk.W, tk.E), pady=(0, 10))
        
        ttk.Label(input_frame, text="Item Name:").grid(row=0, column=0, padx=(0, 5))
        self.item_name_var = tk.StringVar()
        self.item_name_entry = ttk.Entry(input_frame, textvariable=self.item_name_var, width=30)
        self.item_name_entry.grid(row=0, column=1, padx=(0, 10))
        
        ttk.Label(input_frame, text="Price:").grid(row=0, column=2, padx=(0, 5))
        self.price_var = tk.StringVar()
        self.price_entry = ttk.Entry(input_frame, textvariable=self.price_var, width=15)
        self.price_entry.grid(row=0, column=3, padx=(0, 10))
        
        self.create_sale_btn = ttk.Button(
            input_frame, 
            text="Create Sale",
            command=self._on_create_sale
        )
        self.create_sale_btn.grid(row=0, column=4)
        
        # Sales table
        table_frame = ttk.LabelFrame(main_frame, text="Sales History", padding="10")
        table_frame.grid(row=2, column=0, columnspan=2, sticky=(tk.W, tk.E, tk.N, tk.S))
        
        # Configure table frame grid
        table_frame.columnconfigure(0, weight=1)
        table_frame.rowconfigure(0, weight=1)
        
        # Create Treeview with scrollbar
        columns = ('sale_id', 'name', 'original_price', 'discount_percentage', 
                  'discount_amount', 'final_price', 'sale_date')
        
        self.tree = ttk.Treeview(
            table_frame, 
            columns=columns,
            show='headings',
            height=10
        )
        
        # Define headings
        self.tree.heading('sale_id', text='Sale ID')
        self.tree.heading('name', text='Item Name')
        self.tree.heading('original_price', text='Original Price')
        self.tree.heading('discount_percentage', text='Discount %')
        self.tree.heading('discount_amount', text='Discount Amount')
        self.tree.heading('final_price', text='Final Price')
        self.tree.heading('sale_date', text='Sale Date')
        
        # Define columns
        self.tree.column('sale_id', width=120)
        self.tree.column('name', width=150)
        self.tree.column('original_price', width=100)
        self.tree.column('discount_percentage', width=100)
        self.tree.column('discount_amount', width=100)
        self.tree.column('final_price', width=100)
        self.tree.column('sale_date', width=150)
        
        # Add scrollbar
        scrollbar = ttk.Scrollbar(table_frame, orient=tk.VERTICAL, command=self.tree.yview)
        self.tree.configure(yscrollcommand=scrollbar.set)
        
        # Grid tree and scrollbar
        self.tree.grid(row=0, column=0, sticky=(tk.W, tk.E, tk.N, tk.S))
        scrollbar.grid(row=0, column=1, sticky=(tk.N, tk.S))
    
    def set_on_apply_discount_callback(self, callback: Callable[[float], bool]) -> None:
        """Set callback for applying discount."""
        self.on_apply_discount_callback = callback
    
    def set_on_create_sale_callback(self, callback: Callable[[str, float], Any]) -> None:
        """Set callback for creating sale."""
        self.on_create_sale_callback = callback
    
    def _on_apply_discount(self) -> None:
        """Handle apply discount button click."""
        if not self.on_apply_discount_callback:
            return
        
        try:
            discount_value = float(self.discount_var.get())
            success = self.on_apply_discount_callback(discount_value)
            
            if success:
                self.update_current_discount(discount_value)
                messagebox.showinfo("Success", f"Discount updated to {discount_value}%")
            else:
                messagebox.showerror("Error", "Failed to update discount")
        except ValueError:
            messagebox.showerror("Error", "Please enter a valid number for discount")
        except Exception as e:
            messagebox.showerror("Error", f"Failed to apply discount: {str(e)}")
    
    def _on_create_sale(self) -> None:
        """Handle create sale button click."""
        if not self.on_create_sale_callback:
            return
        
        item_name = self.item_name_var.get().strip()
        price_str = self.price_var.get().strip()
        
        if not item_name:
            messagebox.showerror("Error", "Please enter an item name")
            return
        
        if not price_str:
            messagebox.showerror("Error", "Please enter a price")
            return
        
        try:
            price = float(price_str)
            if price <= 0:
                raise ValueError("Price must be greater than zero")
            
            self.on_create_sale_callback(item_name, price)
            
            # Clear input fields
            self.item_name_var.set("")
            self.price_var.set("")
            
        except ValueError as e:
            messagebox.showerror("Error", f"Invalid price: {str(e)}")
    
    def update_sales_table(self, sales_data: List[Dict[str, Any]]) -> None:
        """Update the sales table with new data."""
        # Clear existing items
        for item in self.tree.get_children():
            self.tree.delete(item)
        
        # Add new items
        for sale in sales_data:
            self.tree.insert('', tk.END, values=(
                sale['id'],
                sale['name'],
                f"${sale['original_price']:.2f}",
                f"{sale['discount_percentage']:.1f}%",
                f"${sale['discount_amount']:.2f}",
                f"${sale['final_price']:.2f}",
                sale['sale_date']
            ))
    
    def update_current_discount(self, discount: float) -> None:
        """Update current discount display."""
        self.current_discount_label.config(text=f"Current Discount: {discount:.1f}%")
    
    def show_error(self, message: str) -> None:
        """Show error message."""
        messagebox.showerror("Error", message)

File: hw/cevallos/u3/hw29Singleton/JavaScript/src/views/SalesView.js
class SalesView {
  constructor() {
    this.initializeElements();
    this.setupEventListeners();
    this.initializeView();
  }

  initializeElements() {
    // Form elements
    this.discountInput = document.getElementById('discountPercentage');
    this.applyDiscountBtn = document.getElementById('applyDiscountBtn');
    this.currentDiscountSpan = document.getElementById('currentDiscount');
    
    this.itemNameInput = document.getElementById('itemName');
    this.itemPriceInput = document.getElementById('itemPrice');
    this.createSaleBtn = document.getElementById('createSaleBtn');
    
    // Table elements
    this.salesTableBody = document.getElementById('salesTableBody');
    
    // Statistics elements
    this.totalSalesSpan = document.getElementById('totalSales');
    this.totalRevenueSpan = document.getElementById('totalRevenue');
    this.totalDiscountSpan = document.getElementById('totalDiscount');
    this.averageDiscountSpan = document.getElementById('averageDiscount');
    
    // Message elements
    this.errorMessageDiv = document.getElementById('errorMessage');
    this.successMessageDiv = document.getElementById('successMessage');
    
    // Clear button
    this.clearSalesBtn = document.getElementById('clearSalesBtn');
  }

  setupEventListeners() {
    this.applyDiscountBtn.addEventListener('click', () => this.handleApplyDiscount());
    this.createSaleBtn.addEventListener('click', () => this.handleCreateSale());
    this.clearSalesBtn.addEventListener('click', () => this.handleClearSales());
    
    // Allow Enter key to submit forms
    this.itemNameInput.addEventListener('keypress', (e) => {
      if (e.key === 'Enter') this.handleCreateSale();
    });
    
    this.itemPriceInput.addEventListener('keypress', (e) => {
      if (e.key === 'Enter') this.handleCreateSale();
    });
    
    this.discountInput.addEventListener('keypress', (e) => {
      if (e.key === 'Enter') this.handleApplyDiscount();
    });
  }

  async initializeView() {
    try {
      const currentDiscount = await window.salesAPI.getCurrentDiscount();
      this.updateCurrentDiscount(currentDiscount);
      this.discountInput.value = currentDiscount;
      await this.updateSalesTable();
    } catch (error) {
      this.showMessage(`Error initializing view: ${error.message}`, 'error');
    }
  }

  async handleApplyDiscount() {
    try {
      const discountValue = parseFloat(this.discountInput.value);
      
      if (isNaN(discountValue) || discountValue < 0) {
        throw new Error('Please enter a valid non-negative number for discount.');
      }
      
      if (discountValue > 100) {
        throw new Error('Discount cannot exceed 100%.');
      }
      
      const result = await window.salesAPI.updateDiscount(discountValue);
      
      if (result.success) {
        this.updateCurrentDiscount(discountValue);
        await this.updateSalesTable();
        this.showMessage(`Discount successfully updated to ${discountValue}%`, 'success');
      } else {
        throw new Error(result.error || 'Failed to update discount. Please try again.');
      }
    } catch (error) {
      this.showMessage(error.message, 'error');
    }
  }

  async handleCreateSale() {
    try {
      const itemName = this.itemNameInput.value.trim();
      const priceStr = this.itemPriceInput.value.trim();
      
      if (!itemName) {
        throw new Error('Please enter an item name.');
      }
      
      if (!priceStr) {
        throw new Error('Please enter a price.');
      }
      
      const price = parseFloat(priceStr);
      
      if (isNaN(price) || price <= 0) {
        throw new Error('Price must be a positive number.');
      }
      
      const result = await window.salesAPI.createSale(itemName, price);
      
      if (result.success) {
        // Clear input fields
        this.itemNameInput.value = '';
        this.itemPriceInput.value = '';
        
        // Update UI
        await this.updateSalesTable();
        this.showMessage(
          `Sale created successfully: ${result.sale.name} for $${result.sale.originalPrice.toFixed(2)}`, 
          'success'
        );
        
        // Focus on item name for next entry
        this.itemNameInput.focus();
      } else {
        throw new Error(result.error || 'Failed to create sale.');
      }
    } catch (error) {
      this.showMessage(error.message, 'error');
    }
  }

  async handleClearSales() {
    if (confirm('Are you sure you want to clear all sales? This action cannot be undone.')) {
      try {
        const result = await window.salesAPI.clearSales();
        if (result.success) {
          await this.updateSalesTable();
          this.showMessage('All sales have been cleared.', 'success');
        }
      } catch (error) {
        this.showMessage(`Error clearing sales: ${error.message}`, 'error');
      }
    }
  }

  updateCurrentDiscount(discount) {
    this.currentDiscountSpan.textContent = discount.toFixed(1);
  }

  async updateSalesTable() {
    try {
      const data = await window.salesAPI.getSales();
      const sales = data.sales;
      const stats = data.stats;
      
      // Update table
      this.salesTableBody.innerHTML = '';
      
      if (sales.length === 0) {
        const emptyRow = document.createElement('tr');
        emptyRow.innerHTML = `
          <td colspan="7" style="text-align: center; padding: 2rem; color: #718096;">
            No sales yet. Create your first sale above!
          </td>
        `;
        this.salesTableBody.appendChild(emptyRow);
      } else {
        sales.forEach(sale => {
          const row = document.createElement('tr');
          
          const saleDate = new Date(sale.saleDate);
          const formattedDate = saleDate.toLocaleDateString('en-US', {
            year: 'numeric',
            month: 'short',
            day: 'numeric',
            hour: '2-digit',
            minute: '2-digit'
          });
          
          row.innerHTML = `
            <td>${sale.id}</td>
            <td>${sale.name}</td>
            <td>$${sale.originalPrice.toFixed(2)}</td>
            <td>${sale.discountPercentage.toFixed(1)}%</td>
            <td>$${sale.discountAmount.toFixed(2)}</td>
            <td><strong>$${sale.finalPrice.toFixed(2)}</strong></td>
            <td>${formattedDate}</td>
          `;
          
          this.salesTableBody.appendChild(row);
        });
      }
      
      // Update statistics
      this.updateStatistics(stats);
    } catch (error) {
      this.showMessage(`Error loading sales: ${error.message}`, 'error');
    }
  }

  updateStatistics(stats) {
    this.totalSalesSpan.textContent = stats.totalSales;
    this.totalRevenueSpan.textContent = stats.totalRevenue.toFixed(2);
    this.totalDiscountSpan.textContent = stats.totalDiscount.toFixed(2);
    this.averageDiscountSpan.textContent = stats.averageDiscount.toFixed(2);
  }

  showMessage(message, type = 'error') {
    // Hide both messages first
    this.errorMessageDiv.classList.remove('message-visible');
    this.successMessageDiv.classList.remove('message-visible');
    
    // Show the appropriate message
    if (type === 'error') {
      this.errorMessageDiv.textContent = message;
      this.errorMessageDiv.classList.add('message-visible');
    } else {
      this.successMessageDiv.textContent = message;
      this.successMessageDiv.classList.add('message-visible');
    }
    
    // Auto-hide success messages after 3 seconds
    if (type === 'success') {
      setTimeout(() => {
        this.successMessageDiv.classList.remove('message-visible');
      }, 3000);
    }
  }
}

// Initialize the view when the DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
  window.salesView = new SalesView();
});